include/swift/ABI/Metadata.h

```diff
@@ -3297,6 +3297,40 @@ struct TargetForeignMetadataInitialization {
     CompletionFunction;
 };
 
+/// The cache structure for non-trivial initialization of singleton value
+/// metadata.
+template <typename Runtime>
+struct TargetInPlaceValueMetadataCache {
+  /// The metadata pointer.  Clients can do dependency-ordered loads
+  /// from this, and if they see a non-zero value, it's a Complete
+  /// metadata.
+  std::atomic<TargetMetadataPointer<Runtime, TargetMetadata>> Metadata;
+
+  /// The private cache data.
+  std::atomic<TargetPointer<Runtime, void>> Private;
+};
+using InPlaceValueMetadataCache =
+  TargetInPlaceValueMetadataCache<InProcess>;
+
+/// The control structure for performing non-trivial initialization of
+/// singleton value metadata, which is required when e.g. a non-generic
+/// value type has a resilient component type.
+template <typename Runtime>
+struct TargetInPlaceValueMetadataInitialization {
+  /// The initialization cache.  Out-of-line because mutable.
+  TargetRelativeDirectPointer<Runtime,
+                              TargetInPlaceValueMetadataCache<Runtime>>
+    InitializationCache;
+
+  /// The incomplete metadata.
+  TargetRelativeDirectPointer<Runtime, TargetMetadata<Runtime>>
+    IncompleteMetadata;
+
+  /// The completion function.  The pattern will always be null.
+  TargetRelativeDirectPointer<Runtime, MetadataCompleter>
+    CompletionFunction;
+};
```

Comment:
If you're going to use these for classes, maybe `Value` should be dropped from the class name.

Should the link to the resilient class pattern just be unioned with `IncompleteMetadata`?

Reply:
Unioning the resilient pattern with IncompleteMetadata sounds like a good idea. I'll do that.

I also plan on two follow-up cleanups here:

- renaming 'InPlaceValueMetadata' and similar to 'SingletonMetadata' throughout.
- switching non-generic resilient classes to start from a true-const pattern and not a partial class metadata.

Is it OK if I do the second one in a follow-up PR?

Reply:
Sure.  Do you consider this PR ready for a more complete review?

Reply:
Yes, except I'm redoing the IncompleteMetadata field part.